components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java

```diff
@@ -0,0 +1,370 @@
+            jwtClaimsSet.setAudience(Arrays.asList(request.getApprovedScope()));
+        }
+
+        if (JWSAlgorithm.NONE.getName().equals(signatureAlgorithm.getName())) {
+            return new PlainJWT(jwtClaimsSet).serialize();
+        }
+
+        return signJWT(jwtClaimsSet, null, request);
+    }
+
+    /**
+     * Sign ghe JWT token according to the given signature signing algorithm.
+     * @param jwtClaimsSet JWT claim set to be signed.
+     * @param tokenContext Token context.
+     * @param authorizationContext Authorization context.
+     * @return Signed JWT.
+     * @throws IdentityOAuth2Exception
+     */
+    protected String signJWT(JWTClaimsSet jwtClaimsSet, OAuthTokenReqMessageContext tokenContext,
+                           OAuthAuthzReqMessageContext authorizationContext) throws IdentityOAuth2Exception {
+
+        if (JWSAlgorithm.RS256.equals(signatureAlgorithm) || JWSAlgorithm.RS384.equals(signatureAlgorithm) ||
+                JWSAlgorithm.RS512.equals(signatureAlgorithm)) {
+            return signJWTWithRSA(jwtClaimsSet, tokenContext, authorizationContext);
+        } else if (JWSAlgorithm.HS256.equals(signatureAlgorithm) || JWSAlgorithm.HS384.equals(signatureAlgorithm) ||
+                JWSAlgorithm.HS512.equals(signatureAlgorithm)) {
+            return signJWTWithHMAC(jwtClaimsSet, tokenContext, authorizationContext);
+        } else if (JWSAlgorithm.ES256.equals(signatureAlgorithm) || JWSAlgorithm.ES384.equals(signatureAlgorithm) ||
+                JWSAlgorithm.ES512.equals(signatureAlgorithm)) {
+            return signJWTWithECDSA(jwtClaimsSet, tokenContext, authorizationContext);
+        } else {
+            throw new IdentityOAuth2Exception("Invalid signature algorithm provided. " + signatureAlgorithm);
+        }
+    }
+
+    /**
+     * Sign the JWT token with RSA (SHA-256, SHA-384, SHA-512) algorithm.
+     * @param jwtClaimsSet JWT claim set to be signed.
+     * @param tokenContext Token context if available.
+     * @param authorizationContext Authorization context if available.
+     * @return Signed JWT token.
+     * @throws IdentityOAuth2Exception
+     */
+    protected String signJWTWithRSA(JWTClaimsSet jwtClaimsSet, OAuthTokenReqMessageContext tokenContext,
+                                  OAuthAuthzReqMessageContext authorizationContext) throws IdentityOAuth2Exception {
+
+        try {
+            String tenantDomain = null;
+            if (tokenContext != null) {
+                tenantDomain = tokenContext.getAuthorizedUser().getTenantDomain();
+            } else if (authorizationContext != null) {
+                tenantDomain = authorizationContext.getAuthorizationReqDTO().getUser().getTenantDomain();
+            }
+
+            if (tenantDomain == null) {
+                throw new IdentityOAuth2Exception("Cannot resolve the tenant domain of the user.");
+            }
+
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
```

Comment:
What if someone wants all the access token to be signed with private key of tenant of the App rather than the user?
We have a property to decide this for id_tokens "SignJWTWithSPKey" [1]


[1] https://wso2.org/jira/browse/IDENTITY-6036